DeadHostsStateTests failing on Windows

I see a variety of failures of the following form on the Windows compatibility tests:

```
Suite: org.elasticsearch.client.DeadHostStateTests
FAILURE 0.01s J0 | DeadHostStateTests.testCompareToDefaultTimeSupplier <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: 
   > Expected: a value less than <15864990105023L>
   >      but: <15864990105023L> was equal to <15864990105023L>
   > 	at __randomizedtesting.SeedInfo.seed([CD210BC9F8F37963:56E027485D8503A0]:0)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   > 	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
   > 	at org.elasticsearch.client.DeadHostStateTests.testCompareToDefaultTimeSupplier(DeadHostStateTests.java:67)
   > 	at java.lang.Thread.run(Thread.java:748)
```

E.g. 

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.4+multijob-windows-compatibility/2/console

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.x+multijob-windows-compatibility/6/console

https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-windows-compatibility/3/console

The issue, I believe, is that we use `DeadHostState.TimeSupplier.DEFAULT` in the tests, which wrap `System.getNanoTime()`, and we assume that the value it returns increases between loop iterations. Each iteration likely takes more than 1ns but `System.getNanoTime()` is not so accurate, particularly on Windows, and this causes the tests to fail sometimes.

Marking this as `@AwaitsFix` since I think it'd be better to use our own time supplier throughout these tests rather than `System.getNanoTime()` here, and this seems to require a bit of rework.

--- a/client/rest/src/test/java/org/elasticsearch/client/DeadHostStateTests.java
+++ b/client/rest/src/test/java/org/elasticsearch/client/DeadHostStateTests.java
@@ -20,12 +20,14 @@
 package org.elasticsearch.client;
 
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicLong;
 
 import org.elasticsearch.client.DeadHostState.TimeSupplier;
 
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.greaterThanOrEqualTo;
 import static org.hamcrest.Matchers.is;
 import static org.hamcrest.Matchers.lessThan;
 import static org.junit.Assert.assertEquals;
@@ -36,10 +38,9 @@ public class DeadHostStateTests extends RestClientTestCase {
     private static long[] EXPECTED_TIMEOUTS_SECONDS = new long[]{60, 84, 120, 169, 240, 339, 480, 678, 960, 1357, 1800};
 
     public void testInitialDeadHostStateDefaultTimeSupplier() {
-        assumeFalse("https://github.com/elastic/elasticsearch/issues/33747", System.getProperty("os.name").startsWith("Windows"));
         DeadHostState deadHostState = new DeadHostState(DeadHostState.TimeSupplier.DEFAULT);
         long currentTime = System.nanoTime();
-        assertThat(deadHostState.getDeadUntilNanos(), greaterThan(currentTime));
+        assertThat(deadHostState.getDeadUntilNanos(), greaterThanOrEqualTo(currentTime));
         assertThat(deadHostState.getFailedAttempts(), equalTo(1));
     }
 
@@ -54,13 +55,13 @@ public class DeadHostStateTests extends RestClientTestCase {
         }
     }
 
-    public void testCompareToDefaultTimeSupplier() {
-        assumeFalse("https://github.com/elastic/elasticsearch/issues/33747", System.getProperty("os.name").startsWith("Windows"));
+    public void testCompareToTimeSupplier() {
         int numObjects = randomIntBetween(EXPECTED_TIMEOUTS_SECONDS.length, 30);
         DeadHostState[] deadHostStates = new DeadHostState[numObjects];
         for (int i = 0; i < numObjects; i++) {
             if (i == 0) {
-                deadHostStates[i] = new DeadHostState(DeadHostState.TimeSupplier.DEFAULT);
+                // this test requires a strictly increasing timer
+                deadHostStates[i] = new DeadHostState(new StrictMonotonicTimeSupplier());
             } else {
                 deadHostStates[i] = new DeadHostState(deadHostStates[i - 1]);
             }
@@ -134,6 +135,25 @@ public class DeadHostStateTests extends RestClientTestCase {
         @Override
         public String toString() {
             return "configured[" + nanoTime + "]";
+        }
+    }
+
+    /**
+     * Simulates a monotonically strict increasing time (i.e. the value increases on every call to <code>#nanoTime()</code>). This ensures
+     * that even if we call this time supplier in a very tight loop we always notice time moving forward. This does not happen for real
+     * timer implementations (e.g. on Linux <code>clock_gettime</code> provides microsecond resolution).
+     */
+    static class StrictMonotonicTimeSupplier implements DeadHostState.TimeSupplier {
+        private final AtomicLong time = new AtomicLong(0);
+
+        @Override
+        public long nanoTime() {
+            return time.incrementAndGet();
+        }
+
+        @Override
+        public String toString() {
+            return "strict monotonic[" + time.get() + "]";
         }
     }
 }

